### PR TITLE
Compile with jdk9, targeting jdk8

### PR DESCRIPTION
**Pull request by raboof**
Should result in a jdk8-compatible jar even though the project was compiled
with jdk9, but getting a compilation error related to `sun.reflect.CallerSensitive`

`sbt clean akka-actor/compile` should be sufficient to trigger the problem

### CODE REVIEW

**Comment by raboof** [comment]
Full error:
```
[error] /home/aengelen/dev/akka/akka-actor/src/main/scala/akka/actor/ActorCell.scala:660:9: Class sun.reflect.CallerSensitive not found - continuing with a stub.
[error]         Thread.currentThread().interrupt()
[error]         ^
```

**Comment by raboof** [comment]
I *guess* this happens while parsing the `ct` class file for `Thread`, that has a `@run.reflect.CallerSensitive` annotation on some other methods (though not `currentThread`), and that has probably been [removed in jdk9](https://docs.oracle.com/javase/9/migrate/toc.htm#JSMIG-GUID-B96BD00F-12A4-493A-9907-2FFE8DA6748C).

The message promises it will "continuing with a stub", but it appears to be treated as an error nonetheless. Anything we can do?

**Comment by raboof** [comment]
After adding our own `sun.reflect.CallerSensitive` the `akka-actor` project compiles, but when I try to `publishLocal` I get https://pastebin.com/29LZ69kj while generating scaladoc.

**Comment by ktoso** [comment]
Very weird:

```
[0m[[0m[31merror[0m] [0m[0m     while compiling: /home/aengelen/dev/akka/akka-actor/target/src_managed/main/akka/japi/tuple/Tuples.scala[0m
[0m[[0m[31merror[0m] [0m[0m        during phase: globalPhase=terminal, enteringPhase=typer[0m
[0m[[0m[31merror[0m] [0m[0m     library version: version 2.12.5-bin-85f684d[0m
[0m[[0m[31merror[0m] [0m[0m    compiler version: version 2.12.5-bin-85f684d[0m
```

The file is generated but rather boring: https://gist.github.com/ktoso/199e91aa0ff959e8db0d4733ef3d0ab3

**Comment by ktoso** [comment]
cc @lrytz @retronym -- we're looking into using the `-release` option in this PR, getting weird errors -- see the above one that @raboof pasted

**Comment by retronym** [comment]
Could you gist the stacktace that SBT has suppressed? The SBT command `last` should show it.

**Comment by ktoso** [comment]
Since Arnout sleeps now I'll try to get you that stacktrace, @retronym 👍

**Comment by retronym** [comment]
```
⚡ cat /tmp/test.scala
class Test {
  Thread.currentThread
}

~
⚡ java_use 9; scalac-launch 2.12.5-bin-85f684d -release 8 /tmp/test.scala -d /tmp/foo
~
⚡ java_use 9; scaladoc-launch 2.12.5-bin-85f684d -release 8 /tmp/test.scala -d /tmp/foo
Exception in thread "main" java.lang.AssertionError: assertion failed:
  type AnyRef in java.lang
     while compiling: /tmp/test.scala
        during phase: globalPhase=terminal, enteringPhase=typer
     library version: version 2.12.5-bin-85f684d
    compiler version: version 2.12.5-bin-85f684d
  reconstructed args: -release 8 -usejavacp -d /tmp/foo

  last tree to typer: Select(Select(Select(Ident(java), lang), Thread), currentThread)
       tree position: line 2 of /tmp/test.scala
            tree tpe: ()Thread
              symbol: method currentThread in class Thread
   symbol definition: def currentThread(): Thread (a MethodSymbol)
      symbol package: java.lang
       symbol owners: method currentThread -> class Thread
           call site: class Test in package <empty>

== Source file context for tree position ==

     1 class Test {
     2   Thread.currentThread
     3 }
     4
     5
	at scala.reflect.internal.SymbolTable.throwAssertionError(SymbolTable.scala:157)
	at scala.tools.nsc.doc.model.ModelFactory$modelCreation$.createTemplate(ModelFactory.scala:685)
	at scala.tools.nsc.doc.model.ModelFactory.makeMember0$1(ModelFactory.scala:780)
	at scala.tools.nsc.doc.model.ModelFactory.makeMember(ModelFactory.scala:793)
	at scala.tools.nsc.doc.model.ModelFactory$DocTemplateImpl.$anonfun$ownMembers$1(ModelFactory.scala:353)
	at scala.tools.nsc.doc.model.ModelFactory$DocTemplateImpl.<init>(ModelFactory.scala:353)
	at scala.tools.nsc.doc.model.ModelFactory$PackageImpl.<init>(ModelFactory.scala:462)
	at scala.tools.nsc.doc.model.ModelFactory$modelCreation$$anon$4.<init>(ModelFactory.scala:672)
	at scala.tools.nsc.doc.model.ModelFactory$modelCreation$.createTemplate(ModelFactory.scala:672)
	at scala.tools.nsc.doc.model.ModelFactory.makeMember0$1(ModelFactory.scala:780)
	at scala.tools.nsc.doc.model.ModelFactory.makeMember(ModelFactory.scala:793)
	at scala.tools.nsc.doc.model.ModelFactory$DocTemplateImpl.$anonfun$ownMembers$1(ModelFactory.scala:353)
	at scala.tools.nsc.doc.model.ModelFactory$DocTemplateImpl.<init>(ModelFactory.scala:353)
	at scala.tools.nsc.doc.model.ModelFactory$PackageImpl.<init>(ModelFactory.scala:462)
	at scala.tools.nsc.doc.model.ModelFactory$modelCreation$$anon$4.<init>(ModelFactory.scala:672)
	at scala.tools.nsc.doc.model.ModelFactory$modelCreation$.createTemplate(ModelFactory.scala:672)
	at scala.tools.nsc.doc.model.ModelFactory.makeMember0$1(ModelFactory.scala:780)
	at scala.tools.nsc.doc.model.ModelFactory.makeMember(ModelFactory.scala:793)
	at scala.tools.nsc.doc.model.ModelFactory$DocTemplateImpl.$anonfun$ownMembers$1(ModelFactory.scala:353)
	at scala.tools.nsc.doc.model.ModelFactory$DocTemplateImpl.<init>(ModelFactory.scala:353)
	at scala.tools.nsc.doc.model.ModelFactory$PackageImpl.<init>(ModelFactory.scala:462)
	at scala.tools.nsc.doc.model.ModelFactory$RootPackageImpl.<init>(ModelFactory.scala:470)
	at scala.tools.nsc.doc.model.ModelFactory$modelCreation$$anon$1.<init>(ModelFactory.scala:654)
	at scala.tools.nsc.doc.model.ModelFactory$modelCreation$.createTemplate(ModelFactory.scala:654)
	at scala.tools.nsc.doc.model.ModelFactory$modelCreation$.createRootPackage(ModelFactory.scala:602)
	at scala.tools.nsc.doc.model.ModelFactory$$anon$21.<init>(ModelFactory.scala:45)
	at scala.tools.nsc.doc.model.ModelFactory.makeModel(ModelFactory.scala:42)
	at scala.tools.nsc.doc.DocFactory.makeUniverse(DocFactory.scala:81)
	at scala.tools.nsc.doc.DocFactory.generate$1(DocFactory.scala:124)
	at scala.tools.nsc.doc.DocFactory.document(DocFactory.scala:131)
	at scala.tools.nsc.ScalaDoc.process(ScalaDoc.scala:41)
	at scala.tools.nsc.ScalaDoc$.main(ScalaDoc.scala:92)
	at scala.tools.nsc.ScalaDoc.main(ScalaDoc.scala)
```

**Comment by retronym** [comment]
As a workaround, try removing `-release` from the `scalacOptions in doc`.

**Comment by ktoso** [comment]
Hmm, weird -- I'm seeing `object misc is not a member of package sun`, which is understandable I guess; though a bit weird, since we did in the past after all release already with JDK9... I have to dig into this some more

**Comment by raboof** [comment]
Managed to `publishLocal` akka-actor and akka-remote and verify they indeed no longer have the problem we ran into when publishing akka 2.5.5. 

I haven't seen the error mentioned by @ktoso yet. When running tests I do get `could not find implicit value for parameter pos: org.scalactic.source.Position` errors, looking into that.

**Comment by 2m** [comment]
Added a commit on top of this PR that adds a `CompileJdk9` configuration that compiles sources with `-release 9` flag: https://github.com/2m/akka/commit/06168ea1013489cb60fb4090624f06cac0c8854a

From my initial testing seems to work fine. Needs more smoke tests whether the generated jar is really what we want.

**Comment by lrytz** [comment]
@2m / @ktoso sounds like we can release 2.12.5 then?

**Comment by raboof** [comment]
@lrytz yes we have some nuts to crack on our side but the scala side looks good! Thanks!

**Comment by raboof** [comment]
For future reference:
* putting the classes that should be compiled with different flags in their own sbt `Configuration` seems to work out well, see 2m/akka@f86c163
* `-release 8` also means we cannot use `sun.misc.Unsafe` (http://mail.openjdk.java.net/pipermail/compiler-dev/2016-October/010491.html, http://lists.jboss.org/pipermail/wildfly-dev/2017-January/005692.html). Workarounds could be:
  * Require having jdk8 installed and using `-XDignore.symbol.file -source 8 -target 8 -bootclasspath=$JDK8/rt.jar`
  * Call them via `MethodHandle`/`VarHandle` (though we use a lot of them, so this is inconvenient, https://github.com/scala/collection-strawman/issues/50#issuecomment-344474908)
  * Go through a separate library

**Comment by lrytz** [comment]
Thanks @raboof :)

File: project/CrossJava.scala [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+  def apply(version: String): JavaVersion = CrossJava.parseJavaVersion(version)
+  def apply(numbers: Vector[Long], vendor: String): JavaVersion = new JavaVersion(numbers, Option(vendor))
+}
+
+
+/** Tools for discovering different Java versions,
+  * will be in sbt 1.1.7 (https://github.com/sbt/sbt/pull/4139) but until that time replicated here */
+object CrossJava {
+  object Keys {
+    val discoveredJavaHomes = settingKey[Map[String, File]]("Discovered Java home directories")
```
Comment (patriknw) [inline project/CrossJava.scala]: tried this on my Mac:
```
ListMap(9 -> /Library/Java/JavaVirtualMachines/jdk-9.0.4.jdk/Contents/Home, 1.7 -> /Library/Java/JavaVirtualMachines/jdk1.7.0_45.jdk/Contents/Home, 1.8 -> /Library/Java/JavaVirtualMachines/jdk1.8.0_45.jdk/Contents/Home)
```

However, I also have a newer `jdk1.8.0_121`.
ls /Library/Java/JavaVirtualMachines
```
jdk-9.0.4.jdk		jdk1.7.0_45.jdk		jdk1.8.0.jdk		jdk1.8.0_121.jdk	jdk1.8.0_45.jdk
```

File: project/CrossJava.scala [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    val javaHomes = settingKey[Map[String, File]]("The user-defined additional Java home directories")
+    val fullJavaHomes = settingKey[Map[String, File]]("Combines discoveredJavaHomes and custom javaHomes.")
+  }
+
+  import Keys._
+
+  val crossJavaSettings = Seq(
+    discoveredJavaHomes := CrossJava.discoverJavaHomes,
+    javaHomes := ListMap.empty,
+    fullJavaHomes := CrossJava.expandJavaHomes(discoveredJavaHomes.value ++ javaHomes.value),
```
Comment (patriknw) [inline project/CrossJava.scala]: is the order of importance? should user-defined be first to be able to override?

File: project/AkkaBuild.scala [hunk 3]
```diff
@@ -83,16 +83,26 @@ object AkkaBuild {
 
   private def allWarnings: Boolean = System.getProperty("akka.allwarnings", "false").toBoolean
```
Comment (patriknw) [inline project/AkkaBuild.scala]: I tried `publishLocal` with jdk8 as default `java`.
```
sbt:akka-actor> publishLocal
[info] Packaging /Users/patrik/dev/akka/akka-actor/target/akka-actor_2.11-2.5-SNAPSHOT-sources.jar ...
[info] Done packaging.
[info] Wrote /Users/patrik/dev/akka/akka-actor/target/akka-actor_2.11-2.5-SNAPSHOT.pom
[error] java.util.NoSuchElementException: key not found: 8
[error] 	at scala.collection.immutable.ListMap$Node.applyInternal(ListMap.scala:126)
[error] 	at scala.collection.immutable.ListMap$Node.apply(ListMap.scala:123)
[error] 	at akka.AkkaBuild$.$anonfun$defaultSettings$6(AkkaBuild.scala:104)
```

Comment (patriknw) [reply project/AkkaBuild.scala]: same when using jdk9 as default

File: project/scripts/release [hunk 4]
```diff
@@ -162,7 +162,9 @@ fi
 declare -r version=$1
 declare -r publish_path="${release_server}:${release_path}"
 
-[[ `java -version 2>&1 | grep -E "java version|openjdk version" | cut -d ' ' -f3 | cut -d '.' -f2` -eq 8 ]] || fail "Java version is not 1.8"
+JAVA_VERSION=`java -version 2>&1 | grep -E "java version|openjdk version" | cut -d '"' -f2 | cut -d '.' -f1`
+
+[[ $JAVA_VERSION -ge 9 ]] || fail "Java version is not at least 9"
```
Comment (patriknw) [inline project/scripts/release]: I'm not sure I have an up to date overview picture of what this is doing. It requires jdk9 when releasing, and will also use jdk8 rt.jar for some parts? You could explain in today's meeting, but perhaps worth documenting in some comment how it works.

**Review by patriknw — COMMENTED** [review]
tried on osx and perhaps found some remaining issues...

File: project/CrossJava.scala (hunk 1 above)
Comment (raboof) [reply project/CrossJava.scala]: Ow, for me on Debian it is:
```
ListMap(9 -> /usr/lib/jvm/java-9-openjdk-amd64, 8 -> /usr/lib/jvm/java-8-openjdk-amd64, 10 -> /usr/lib/jvm/java-10-openjdk-amd64)
```
so notably for me the key is '8', not '1.8'.

As noted in the comment I temporarily took this code from SBT. @2m do you know what the desired format was? I guess that should be fixed both here and in SBT?

File: project/CrossJava.scala (hunk 2 above)
Comment (raboof) [reply project/CrossJava.scala]: @2m WDYT?

File: project/CrossJava.scala (hunk 1 above)
Comment (2m) [reply project/CrossJava.scala]: It should actually accept both: https://github.com/sbt/sbt/commit/aff9e0110cd7f4b942e194e1eb979f89fbf64f1f#diff-d0af389abdaf608fbd5bbd98420b2e27

File: project/CrossJava.scala (hunk 2 above)
Comment (2m) [reply project/CrossJava.scala]: Yes, user-defined ones should override the discovered ones.

File: project/CrossJava.scala (hunk 1 above)
Comment (raboof) [reply project/CrossJava.scala]: Ah, but that only goes for `fullJavaHomes`, not `discoveredJavaHomes` - I should've used `fullJavaHomes` in any case I suppose.

Comment (2m) [reply project/CrossJava.scala]: Yup. `fullJavaHomes` is the discovered plus user overrides. Similar to `resolvers` and `fullResolvers`.

Comment (patriknw) [reply project/CrossJava.scala]: It's still not finding my newer `jdk1.8.0_121`

```
sbt:akka-actor> show discoveredJavaHomes
[info] ListMap(9 -> /Library/Java/JavaVirtualMachines/jdk-9.0.4.jdk/Contents/Home, 1.7 -> /Library/Java/JavaVirtualMachines/jdk1.7.0_45.jdk/Contents/Home, 1.8 -> /Library/Java/JavaVirtualMachines/jdk1.8.0_45.jdk/Contents/Home)

sbt:akka-actor> show fullJavaHomes
[info] Map(1.9 -> /Library/Java/JavaVirtualMachines/jdk-9.0.4.jdk/Contents/Home, 8 -> /Library/Java/JavaVirtualMachines/jdk1.8.0_45.jdk/Contents/Home, 9 -> /Library/Java/JavaVirtualMachines/jdk-9.0.4.jdk/Contents/Home, 1.7 -> /Library/Java/JavaVirtualMachines/jdk1.7.0_45.jdk/Contents/Home, 1.8 -> /Library/Java/JavaVirtualMachines/jdk1.8.0_45.jdk/Contents/Home, 7 -> /Library/Java/JavaVirtualMachines/jdk1.7.0_45.jdk/Contents/Home)
```

```
ls /Library/Java/JavaVirtualMachines
jdk-9.0.4.jdk		jdk1.7.0_45.jdk		jdk1.8.0.jdk		jdk1.8.0_121.jdk	jdk1.8.0_45.jdk
```

File: project/AkkaBuild.scala (hunk 3 above)
Comment (patriknw) [reply project/AkkaBuild.scala]: confirmed, this works for me now

File: project/CrossJava.scala (hunk 1 above)
Comment (raboof) [reply project/CrossJava.scala]: Ah yes, I'll look into that one. Looks like it just picks a random one right now.

Comment (raboof) [reply project/CrossJava.scala]: @patriknw should we prefer jdk1.8.0_121.jdk to jdk1.8.0.jdk Not sure if the latter is a symlink or more like `jdk1.8.0_0.jdk`?

Comment (patriknw) [reply project/CrossJava.scala]: no symlink, `jdk1.8.0.jdk` is a really old one. I'm fine with removing that one and you can assume that they have a `_version` that can be sorted

**Review by patriknw — APPROVED** [review]
LGTM

**Review by johanandren — APPROVED** [review]
Not sure what the status is but I think it LGTM

**Comment by raboof** [comment]
Should be good to go!

**Pull request merged**